cfgmgr/sflowmgr.cpp

```diff
@@ -0,0 +1,370 @@
+                port_info.local_rate = "";
+                port_info.local_admin = "";
+                m_sflowPortLocalConfMap[key] = port_info;
+            }
+            for (auto i : values)
+            {
+                if (fvField(i) == "speed")
+                {
+                    m_sflowPortLocalConfMap[key].speed = fvValue(i);
+                }
+            }
+
+            if (new_port)
+            {
+                if (gEnable && intf_all_conf)
+                {
+                    vector<FieldValueTuple> fvs;
+                    sflowGetGlobalFvs(fvs, m_sflowPortLocalConfMap[key].speed);
+                    m_appSflowSessionTable.set(key, fvs);
+                }
+            }
+        }
+        else if (op == DEL_COMMAND)
+        {
+            auto sflowPortConf = m_sflowPortLocalConfMap.find(key);
+            if (sflowPortConf != m_sflowPortLocalConfMap.end())
+            {
+                bool local_cfg = m_sflowPortLocalConfMap[key].local_conf;
+
+                m_sflowPortLocalConfMap.erase(key);
+                if ((intf_all_conf && gEnable) || local_cfg)
+                {
+                    m_appSflowSessionTable.del(key);
+                }
+            }
+        }
+        it = consumer.m_toSync.erase(it);
+    }
+}
+
+void SflowMgr::sflowHandleSessionAll(bool enable)
+{
+    for (auto it: m_sflowPortLocalConfMap)
+    {
+        if (!it.second.local_conf)
+        {
+            vector<FieldValueTuple> fvs;
+            sflowGetGlobalFvs(fvs, it.second.speed);
+            if (enable)
+            {
+                m_appSflowSessionTable.set(it.first, fvs);
+            }
+            else
+            {
+                m_appSflowSessionTable.del(it.first);
+            }
+        }
+    }
+}
+
+void SflowMgr::sflowHandleSessionLocal(bool enable)
+{
+    for (auto it: m_sflowPortLocalConfMap)
+    {
+        if (it.second.local_conf)
+        {
+            vector<FieldValueTuple> fvs;
+            sflowGetLocalFvs(fvs, it.second);
+            if (enable)
+            {
+                m_appSflowSessionTable.set(it.first, fvs);
+            }
+            else
+            {
+                m_appSflowSessionTable.del(it.first);
+            }
+        }
+    }
+}
+
+void SflowMgr::sflowGetGlobalFvs(vector<FieldValueTuple> &fvs, string speed)
+{
+    string rate;
+    FieldValueTuple fv1("admin_state", "enable");
+    fvs.push_back(fv1);
+
+    if (speed != SFLOW_ERROR_SPEED_STR)
+    {
+        rate = sflowSpeedRateInitMap[speed];
+    }
+    else
+    {
+        rate = SFLOW_ERROR_SPEED_STR;
+    }
+    FieldValueTuple fv2("sample_rate",rate);
+    fvs.push_back(fv2);
+}
+
+void SflowMgr::sflowGetLocalFvs(vector<FieldValueTuple> &fvs, SflowLocalPortInfo &local_info)
+{
+    if (local_info.local_admin.length() > 0)
+    {
+        FieldValueTuple fv1("admin_state", local_info.local_admin);
+        fvs.push_back(fv1);
+    }
+
+    FieldValueTuple fv2("sample_rate", local_info.local_rate);
```

Comment:
Hi Prince,
While it is not required for the local_admin to be externally configured, the default can be derived in orchagent from "sflow status" using config sflow enable. So only when admin is configured externally for a port that overrides global setting, this field needs to be filled.
For rate however, the manager must necessarily fill the information at all cases. If externally configured the rate string is used as such and if not the rate is derived from the speed of the interface. Thus the rate string will always exists while admin is only when configured which is the reason why the check is not done for rate

Reply:
I have modified the logic so that both admin as well as rate are published and both information would be present in app-db even if user CLI provides only one. The other would be derived and published in appDb so that both orchagent and CLI would have clear picture of the configuration rather than having assumptions